### PR TITLE
Flag Warnings as Errors during Documentation Build

### LINKED ISSUE #9778: Documentation build in CI does not flag some kinds of errors
### Is your proposal related to a problem?

When I am editing the documentation and I build the docs locally I often see that there are missing images or references that somehow are not being flagged at the CI level.

### Describe the solution you'd like

All warnings should fail the CI build for documentation.

### Describe alternatives you've considered

* We somehow change these warnings to errors (which may fail the CI)
* We just ignore these and hope that developers see these and fix these up manually

### Additional context

Example output from December 2022 - note that there are missing images and missing links (many happened when we moved the guide out).

* /code/wagtail/docs/advanced_topics/add_to_django_project.md:98: WARNING: 'myst' reference target not found: inserting_videos
* /code/wagtail/docs/advanced_topics/deploying.md:70: WARNING: 'myst' reference target not found: collection_privacy_settings
* /code/wagtail/docs/reference/contrib/redirects.md:36: WARNING: 'myst' reference target not found: managing_redirects
* /code/wagtail/docs/reference/pages/panels.md:1: WARNING: undefined label: 'commenting'

Full stack

```
root@e4e1af4133fa:/code/wagtail# cd docs/
root@e4e1af4133fa:/code/wagtail/docs# make html
sphinx-build -b html -d _build/doctrees   . _build/html
Running Sphinx v5.3.0
Initializing Spelling Checker 5.4.0
making output directory... done
loading intersphinx inventory from https://docs.djangoproject.com/en/stable/_objects/...
intersphinx inventory has moved: https://docs.djangoproject.com/en/stable/_objects/ -> https://docs.djangoproject.com/en/4.1/_objects/
myst v0.18.1: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=[], disable_syntax=[], all_links_external=False, url_schemes=('http', 'https', 'mailto', 'ftp'), ref_domains=None, highlight_code_blocks=True, number_code_blocks=[], title_to_header=False, heading_anchors=None, heading_slug_func=None, footnote_transition=True, words_per_minute=200, sub_delimiters=('{', '}'), linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area')
building [mo]: targets for 0 po files that are out of date
building [html]: targets for 266 source files that are out of date
updating environment: [new config] 266 added, 0 changed, 0 removed
reading sources... [100%] topics/writing_templates                                                                                                     
/code/wagtail/docs/extending/extending_draftail.md:152: WARNING: image file not readable: _static/images/draftail_entity_stock_source.gif
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
writing output... [100%] topics/writing_templates                                                                                                      
/code/wagtail/docs/advanced_topics/add_to_django_project.md:98: WARNING: 'myst' reference target not found: inserting_videos
/code/wagtail/docs/advanced_topics/deploying.md:70: WARNING: 'myst' reference target not found: collection_privacy_settings
/code/wagtail/docs/reference/contrib/redirects.md:36: WARNING: 'myst' reference target not found: managing_redirects
/code/wagtail/docs/reference/pages/panels.md:1: WARNING: undefined label: 'commenting'
/code/wagtail/docs/reference/settings.md:386: WARNING: 'myst' reference target not found: collection_privacy_settings
/code/wagtail/docs/releases/2.10.rst:20: WARNING: undefined label: 'managing_workflows'
/code/wagtail/docs/releases/2.10.rst:154: WARNING: undefined label: 'managing_workflows'
/code/wagtail/docs/releases/2.11.rst:142: WARNING: undefined label: 'ie11'
/code/wagtail/docs/releases/2.13.rst:34: WARNING: undefined label: 'commenting'
/code/wagtail/docs/releases/4.2.md:72: WARNING: 'myst' reference target not found: performance
/code/wagtail/docs/topics/snippets.md:413: WARNING: more than one target found for 'myst' cross-reference permissions: could be :std:ref:`Permissions` or :std:doc:`Permissions` [myst.ref]
/code/wagtail/docs/topics/writing_templates.md:63: WARNING: 'myst' reference target not found: inserting_images
generating indices... py-modindex done
writing additional pages... search done
copying images... [100%] _static/images/image_filter_fill_focal_close.png                                                                              
copying static files... done
copying extra files... done
dumping search index in English (code: en)... done
dumping object inventory... done
build succeeded, 13 warnings.

The HTML pages are in _build/html.

Build finished. The HTML pages are in _build/html.
```

### PR DESCRIPTION
**Pull request by NXPY123**
Fixes #9778 warnings during documentation build will now be errors and build will not be completed until they are resolved.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [ ] For new features: Has the documentation been updated accordingly?

### CODE REVIEW

**Comment by lb-** [comment]
@NXPY123 could you maybe push up a new commit that's intentionally going to fail the CI (for testing).

Curiously, when I run the build locally I am still getting these two errors below. These are false negatives (due to the code snippet being incorrect HTML but actually what we used to do with some HTML). and I have not worked out how to ignore them (maybe MyST allows ignoring somehow).

```
/code/wagtail/docs/releases/5.2.md:502: WARNING: Lexing literal_block '{% load wagtailadmin_tags %}\n<script type="text/django-form-template" id="id_{{ formset.prefix }}-EMPTY_FORM_TEMPLATE">\n    {% escapescript %}\n        <div>Widget template content</div>\n        <script src="/js/my-widget.js"></script>\n    {% endescapescript %}\n</script>\n' as "html+django" resulted in an error at token: '/'. Retrying in relaxed mode.

/code/wagtail/docs/releases/6.0.md:297: WARNING: Lexing literal_block '<form\n  method="POST"\n  data-controller="w-unsaved"\n  data-action="w-unsaved#submit beforeunload@window->w-unsaved#confirm change->w-unsaved#check keyup->w-unsaved#check"\n  data-w-unsaved-confirm-value="This page has unsaved changes." // equivalent to `confirmationMessage`.\n  data-w-unsaved-force-value="true" // equivalent to `alwaysDirty`.\n  data-w-unsaved-watch-value="edits comments" // can add \'comments\' if comments is enabled, defaults to only \'edits\'.\n  >\n  ... form contents\n</form>\n' as "javascript" resulted in an error at token: '/'. Retrying in relaxed mode.
```

We could update those two snippets to be `text` instead of `html+django`.

But the point is, I thought these two items would be flagged if the CI is checking all docs, maybe it's only checking changed files?

**Comment by NXPY123** [comment]
I'm sorry about that. I didn't check for existing warnings when I made the change. When I looked through the [docs](https://www.sphinx-doc.org/en/master/man/sphinx-build.html) of `sphinx-build`, it suggested using `-E` to completely rebuild the docs each time. This will help ensure all warnings are detected and the build will exit with Error 1 if there are any but the time to build on each call would be high.

**Comment by lb-** [comment]
@NXPY123 ok. That makes sense. I think it's best to keep the current partial rebuild if that's the current approach. 

It would be still good to push a dummy commit to this branch that you know should fail. This way we can validate the CI. 

Then if it fails - we know it's working - after that you can remove the commit.

**Comment by laymonage** [comment]
I'd be fine if `make html` rebuilds the whole thing, as long as we keep the partial rebuild for `make livehtml`. Alternatively, perhaps introduce a new `make` recipe or change it so that you can pass `-E` manually (I doubt you can do that with `make` without turning it into an environment variable, though).

**Comment by gasman** [comment]
I'd rather not change the behaviour of `make html` if we can help it. From experience, it seems to do a full rebuild if you haven't built docs for a while, and the long build time is already pretty annoying - having to remember to use `make livehtml` up-front to make quick iterative changes is going to be more annoying.

**Comment by lb-** [comment]
Just a note. The goal here is to only change the output for CI builds so that an error in the build will fail the CI. 

Ideally we would not change the behaviour of `make html` locally and also not impact current performance on the CI. 

Use case. A new documentation update has a malformed link that will not work at all, the CI should not pass this (as it currently does).

**Comment by NXPY123** [comment]
I updated the `.readthedocs.yml` to set `fail_on_warning` under the `sphinx` section to `True` and the build failed for a warning.

Build Log
```
Running Sphinx v7.2.6
making output directory... done
loading intersphinx inventory from https://docs.djangoproject.com/en/stable/_objects/...
intersphinx inventory has moved: https://docs.djangoproject.com/en/stable/_objects/ -> https://docs.djangoproject.com/en/5.0/_objects/
myst v2.0.0: MdParserConfig(commonmark_only=False, gfm_only=False, enable_extensions=set(), disable_syntax=[], all_links_external=False, url_schemes=('https', 'http', 'mailto'), ref_domains=None, fence_as_directive=set(), number_code_blocks=[], title_to_header=False, heading_anchors=0, heading_slug_func=None, html_meta={}, footnote_transition=True, words_per_minute=200, substitutions={}, linkify_fuzzy_links=True, dmath_allow_labels=True, dmath_allow_space=True, dmath_allow_digits=True, dmath_double_inline=False, update_mathjax=True, mathjax_classes='tex2jax_process|mathjax_process|math|output_area', enable_checkboxes=False, suppress_warnings=[], highlight_code_blocks=True)
building [mo]: targets for 0 po files that are out of date
writing output... 
building [html]: targets for 304 source files that are out of date
updating environment: [new config] 304 added, 0 changed, 0 removed
[2Kreading sources... [  0%] advanced_topics/accessibility_considerations
[2Kreading sources... [100%] tutorial/set_up_site_menu
[2Kreading sources... [100%] tutorial/style_your_site
/home/docs/checkouts/readthedocs.org/user_builds/wagtail/checkouts/11503/docs/advanced_topics/deploying.md:17: WARNING: No matches for 'djafngo:*:*:howto/deployment/wsgi/index' [myst.iref_missing]
looking for now-outdated files... none found
pickling environment... done
checking consistency... done
preparing documents... done
copying assets... copying static files... done
copying extra files... done
done
[2Kwriting output... [  0%] advanced_topics/accessibility_considerations
[2Kwriting output... [100%] tutorial/set_up_site_menu
[2Kwriting output... [100%] tutorial/style_your_site
/home/docs/checkouts/readthedocs.org/user_builds/wagtail/checkouts/11503/docs/releases/5.2.md:502: WARNING: Lexing literal_block '{% load wagtailadmin_tags %}\n<script type="text/django-form-template" id="id_{{ formset.prefix }}-EMPTY_FORM_TEMPLATE">\n    {% escapescript %}\n        <div>Widget template content</div>\n        <script src="/js/my-widget.js"></script>\n    {% endescapescript %}\n</script>\n' as "html+django" resulted in an error at token: '/'. Retrying in relaxed mode.
/home/docs/checkouts/readthedocs.org/user_builds/wagtail/checkouts/11503/docs/releases/6.0.md:297: WARNING: Lexing literal_block '<form\n  method="POST"\n  data-controller="w-unsaved"\n  data-action="w-unsaved#submit beforeunload@window->w-unsaved#confirm change->w-unsaved#check keyup->w-unsaved#check"\n  data-w-unsaved-confirm-value="This page has unsaved changes." // equivalent to `confirmationMessage`.\n  data-w-unsaved-force-value="true" // equivalent to `alwaysDirty`.\n  data-w-unsaved-watch-value="edits comments" // can add \'comments\' if comments is enabled, defaults to only \'edits\'.\n  >\n  ... form contents\n</form>\n' as "javascript" resulted in an error at token: '/'. Retrying in relaxed mode.
generating indices... py-modindex done
writing additional pages... search done
[2Kcopying images... [  3%] _static/images/screen40_table_block.png
[2Kcopying images... [100%] _static/images/image_filter_fill_focal_close.png
dumping search index in English (code: en)... done
dumping object inventory... done
Updating searchtools for Read the Docs search... build finished with problems, 3 warnings.
```

**Comment by lb-** [comment]
Ok. Those logs are a bit hard to parse but this looks like progress. 

Next step will be to get the build passing by updating the two failed html code snippets. 

As noted above, we can just change them to text.

**Comment by lb-** [comment]
I have just rebased and pushed this with one more fix. Will see how the CI goes on the latest `main`.

Thanks @NXPY123  - I think this is close.
